### PR TITLE
Advance deprecation of opDot to error

### PR DESCRIPTION
**Pull request by JinShil**
Followup to https://github.com/dlang/dmd/pull/8406

I'm expecting this to be merged after 2.087 is released.

### CODE REVIEW

File: src/dmd/typesem.d [hunk 1]
```diff
@@ -3384,14 +3384,12 @@ Expression dotExp(Type mt, Scope* sc, Expression e, Identifier ident, int flag)
...
-                e = build_overload(e.loc, sc, e, null, fd);
-                // @@@DEPRECATED_2.087@@@.
-                e.deprecation("`opDot` is deprecated. Use `alias this`");
-                e = new DotIdExp(e.loc, e, ident);
-                return returnExp(e.expressionSemantic(sc));
+                // @@@DEPRECATED_2.094@@@.
+                // Deprecated in 2.082
+                // Made an error in 2.088
+                // Remove in 2.094
+                e.error("`opDot` is deprecated. Use `alias this`");
```
Comment (Geod24) [inline src/dmd/typesem.d]: `s/is deprecated/had been removed/`

**Comment by rainers** [comment]
Let me note that I had a terrible experience when using `alias this` instead of `opDot` in https://github.com/rainers/visuald/commit/9f93772da2814aa16cbac738710e12f120bc5455 which silently produced crashing code. I had to replace it with `opDispatch` in https://github.com/rainers/visuald/commit/ead3f510b207ae909d74c68f2cd75a160d6cd114.

This was happening in combination with other operator overloads. Was the order of evaluation changed recently to fix that? If not, should this workaround be mentioned in the changelog, too?

**Comment by Basile-z** [comment]
@rainers, is there a minimized test case for the problem you encountered in VD ?

**Comment by rainers** [comment]
> @rainers, is there a minimized test case for the problem you encountered in VD ?

Not yet, I can try to reduce the one I encountered.

**Comment by JinShil** [comment]
> Was the order of evaluation changed recently to fix that?

It may have been addressed with https://github.com/dlang/dmd/pull/9289.

**Comment by rainers** [comment]
This is a partly reduced test case that fails with `-version=aliasThis`, but not without:
```
// simple double linked list

struct DListNode
{
	int data;

	DListNode* _next;
	DListNode* _prev;
}

class DList
{
	DListNode root;

	this()
	{
		root._next = &root;
		root._prev = &root;
	}

	void append(int item)
	{
		DListNode *node = new DListNode;
		node.data = item;

		node._next = &root;
		node._prev = root._prev;
		root._prev._next = node;
		root._prev = node;
	}

	DListIterator begin()
	{
		DListIterator it = DListIterator(this);
		it._pos = root._next;
		return it;
	}

	DListIterator end()
	{
		DListIterator it = DListIterator(this);
		it._pos = &root;
		return it;
	}

	ref int opIndex(int idx)
	{
		DListIterator it = begin();
		it += idx;
		return *it;
	}
}

struct DListIterator
{
	DListNode* _pos;
	DList _list;

	this(DList list)
	{
		_list = list;
	}

version(aliasThis) // opDot deprecated
{
	alias opStar this;

	// https://issues.dlang.org/show_bug.cgi?id=16657 need to define opEquals with alias this
	bool opEquals(const typeof(this) other) const
	{
		return _pos == other._pos; // no need to compare lists, nodes belong to only one list
	}
}
else // opDot
{
	ref int opDot()
	{
		return _pos.data;
	}
}

	ref int opStar()
	{
		return _pos.data;
	}

	ref int opIndex(int idx)
	{
		DListIterator it = this;
		it += idx;
		return *it;
	}

	DListIterator opAdd(int cnt)
	{
		DListIterator it = this;
		it += cnt;
		return it;
	}

	DListIterator opSub(int cnt)
	{
		DListIterator it = this;
		it += -cnt;
		return it;
	}

	void opAddAssign(int cnt)
	{
		foreach(_; 0..cnt)
			_pos = _pos._next;
	}
}

unittest
{
	DList list = new DList;

	list.append(1);
	list.append(2);

	DListIterator it = list.begin();
	assert(*it == 1);
	++it;                  // with alias this, this modifies the data, not the iterator
	assert(*it == 3); // fails with alias this
	++it;
	assert(it == list.end());
}
```

The opDot operator isn't used here, though, as the aliased type is an int.

**Comment by rainers** [comment]
> It may have been addressed with #9289.

I suspect that only affects comparisons, not other operators.

**Comment by JinShil** [comment]
@rainers Please see https://issues.dlang.org/show_bug.cgi?id=18665

It is my understanding that `opStar`, `opSub`, `opAdd`, etc.. should all be deprecated in favor of `opUnary`.  In fact those operator overloads are not even documented anymore at https://dlang.org/spec/operatoroverloading.html.  Can you refactor your example to use `opUnary` and avoid the issue?

**Comment by Basile-z** [comment]
Yes there are all D1 operators. In the absence of specification there's no way to say if things are working correctly or not. IIRC In opover.d D1-style overloads are tried as a fallback so D2-style win.

**Comment by Basile-z** [comment]
I mean that the effect of blending the two styles is not defined per spec, but it has not to.

**Comment by rainers** [comment]
Indeed, using opUnary seems to work.

**Comment by Geod24** [comment]
Has any thought been given to deprecating those operators ?

@JinShil : Also note that this breaks fairly recent versions of Vibe.d, which is not something we should be keen on, considering Vibe.d is probably the most popular library out there.

**Comment by JinShil** [comment]
> Has any thought been given to deprecating those operators ?

Yes, I'd like to do that, but I have to do this kind of work incrementally.  First is to move the `opDot` deprecation forward.  Next I'll submit PRs to deprecate the other D1 operators, assuming I can figure out how to do it.

> Also note that this breaks fairly recent versions of Vibe.d, which is not something we should be keen on, considering Vibe.d is probably the most popular library out there.

I certainly don't want to cause disruption, but I'd like to remove technical debt and unnecessary complexity in the language and the compiler.  It seems the primary vibe.d tests pass, but some packages using older versions of vibe.d fail.  I've never used vibe.d so I don't know how serious this is.  How recent is "fairly recent"?

**Comment by Geod24** [comment]
> I certainly don't want to cause disruption, but I'd like to remove technical debt and unnecessary complexity in the language and the compiler. It seems the primary vibe.d tests pass, but some packages using older versions of vibe.d fail. I've never used vibe.d so I don't know how serious this is. How recent is "fairly recent"?

To put it midly, Vibe.d has been doing a stellar job at testing with all compiler. [Just look at Travis config on master](https://github.com/vibe-d/vibe.d/blob/665ea90c80b8e3d75af41c66b4d01764a6b98ddb/.travis.yml#L21-L46). Most of the time I can just come back to a project 2 years later and have very little modifications to do to make it compile and running.
From a quick look, the failing job depends on vibe-core 1.4.1 which is almost 1 year old (released July 9). 1.4.2 was released in early September 2108.

> Yes, I'd like to do that, but I have to do this kind of work incrementally. First I would like to move the `opDot` deprecation forward. Next I'll submit PRs to deprecate the other D1 operators, assuming I can figure out how to do it.

In @rainers ' case, it looks like deprecating all the operators would have avoided the issue he went through. I just thought it's reasonable to assume that any code base using `opDot` might have other D1-style operators, and thus run into the same issue. So if we deprecate those operators now, we might save them the trouble.

**Comment by JinShil** [comment]
> From a quick look, the failing job depends on vibe-core 1.4.1 which is almost 1 year old (released July 9). 1.4.2 was released in early September 2018.

What do you recommend?

**Comment by wilzbach** [comment]
@Geod24: this can only be merged if there are latest versions of vibe-core etc on the project tester that don't use opDot anymore. We have never required anything else and done so plenty of times. The problem with Vibe.d is that it has almost no maintainance and thus the deprecation messages were ignored.

**Comment by Geod24** [comment]
@JinShil : Since we now have a release where D1 operators are deprecated, I think we can proceed with this ?

**Comment by JinShil** [comment]
I've rebased this, but it can't move forward until the failing projects in Buildkite are updated to a later version of Vibe.

**Pull request closed without merging**